dll/cpl/hotplug/enum.c

```diff
@@ -4,4 +4,308 @@
+        if (cr != CR_SUCCESS)
+        {
+            //DPRINT("No sibling! %lu\n", cr);
+            return;
+        }
+
+        hTreeItem = InsertDeviceTreeItem(hParentItem,
+                                         ChildDevInst,
+                                         pHotplugData);
+        if (hTreeItem != NULL)
+        {
+            DevTreeRecursiveInsertSubDevices(hTreeItem,
+                                             ChildDevInst,
+                                             pHotplugData);
+        }
+    }
+}
+
+VOID
+EnumHotpluggedDevices(
+    _In_ PHOTPLUG_DATA pHotplugData)
+{
+    SP_DEVINFO_DATA did = { 0 };
+    HDEVINFO hdev;
+    int idev;
+    DWORD dwCapabilities, dwSize;
+    ULONG ulStatus, ulProblem;
+    HTREEITEM hTreeItem;
+    CONFIGRET cr;
+
+    //DPRINT1("EnumHotpluggedDevices()\n");
+
+    TreeView_DeleteAllItems(pHotplugData->hwndDeviceTree);
+
+    hdev = SetupDiGetClassDevs(NULL, NULL, 0, DIGCF_ALLCLASSES | DIGCF_PRESENT);
+    if (hdev == INVALID_HANDLE_VALUE)
+        return;
+
+    did.cbSize = sizeof(did);
+
+    /* Enumerate all the attached devices */
+    for (idev = 0; SetupDiEnumDeviceInfo(hdev, idev, &did); idev++)
+    {
+        ulStatus = 0;
+        ulProblem = 0;
+
+        cr = CM_Get_DevNode_Status(&ulStatus,
+                                   &ulProblem,
+                                   did.DevInst,
+                                   0);
+        if (cr != CR_SUCCESS)
+            continue;
+
+        dwCapabilities = 0,
+        dwSize = sizeof(dwCapabilities);
+        cr = CM_Get_DevNode_Registry_Property(did.DevInst,
+                                              CM_DRP_CAPABILITIES,
+                                              NULL,
+                                              &dwCapabilities,
+                                              &dwSize,
+                                              0);
+        if (cr != CR_SUCCESS)
+            continue;
+
+        /* Add devices that require safe removal to the device tree */
+        if ( (dwCapabilities & CM_DEVCAP_REMOVABLE) &&
+            !(dwCapabilities & CM_DEVCAP_DOCKDEVICE) &&
+            !(dwCapabilities & CM_DEVCAP_SURPRISEREMOVALOK) &&
+            ((dwCapabilities & CM_DEVCAP_EJECTSUPPORTED) || (ulStatus & DN_DISABLEABLE)) &&
+            ulProblem == 0)
+        {
+            hTreeItem = InsertDeviceTreeItem(TVI_ROOT,
+                                             did.DevInst,
+                                             pHotplugData);
+
+            if ((hTreeItem != NULL) && (pHotplugData->dwFlags & HOTPLUG_DISPLAY_DEVICE_COMPONENTS))
+            {
+                DevTreeRecursiveInsertSubDevices(hTreeItem,
+                                                 did.DevInst,
+                                                 pHotplugData);
+            }
+        }
+    }
+
+    SetupDiDestroyDeviceInfoList(hdev);
+}
+
+static
+VOID
+InsertConfirmDeviceListItem(
```

Comment:
This function appears to be identical with the InsertDeviceTreeItem one, except for the very last part (`/* Add it to the confirm device list */` ...) . Thus I was wondering whether you coud, for reducing code duplication, define a common function, and then these two Insert...Item functions would call that common function, then each do the different kind of insertion (insertion either in treeview, or in listview).

That hypothetical common function (that would be called by the two different Insert...Item helpers), could take in parameters: the `_In_ DEVINST DevInst`, the `_In_ PHOTPLUG_DATA pHotplugData`, and return, via parameters, the corresponding displayName, and the nClassImage (via `_Out_` pointers).

Reply:
Fixed.